File: Graphs/BiconnectedGraph.py
# File Name: BiconnectedGraph.py

from collections import defaultdict

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = defaultdict(list)
        self.time = 0

    def addEdge(self, u, v):
        self.graph[u].append(v)
        self.graph[v].append(u)

    def APUtil(self, u, visited, parent, low, disc, ap):
        children = 0
        visited[u] = True
        disc[u] = low[u] = self.time
        self.time += 1

        for v in self.graph[u]:
            if not visited[v]:
                parent[v] = u
                children += 1
                self.APUtil(v, visited, parent, low, disc, ap)

                low[u] = min(low[u], low[v])

                # Root case
                if parent[u] == -1 and children > 1:
                    ap[u] = True

                # Non-root case
                if parent[u] != -1 and low[v] >= disc[u]:
                    ap[u] = True

            elif v != parent[u]:
                low[u] = min(low[u], disc[v])

    def findArticulationPoints(self):
        visited = [False] * self.V
        parent = [-1] * self.V
        low = [float("inf")] * self.V
        disc = [float("inf")] * self.V
        ap = [False] * self.V

        for i in range(self.V):
            if not visited[i]:
                self.APUtil(i, visited, parent, low, disc, ap)

        return ap

    def isConnected(self):
        visited = [False] * self.V
        self.DFS(0, visited)
        return all(visited)

    def DFS(self, v, visited):
        visited[v] = True
        for neighbour in self.graph[v]:
            if not visited[neighbour]:
                self.DFS(neighbour, visited)

    def countLeaves(self):
        count = 0
        for i in range(self.V):
            if len(self.graph[i]) == 1:
                count += 1
        return count


# -------- Driver Code --------
if __name__ == "__main__":

    V = int(input("Enter number of vertices: "))
    E = int(input("Enter number of edges: "))

    g = Graph(V)

    print("Enter edges (u v):")
    for _ in range(E):
        u, v = map(int, input().split())
        g.addEdge(u, v)

    ap = g.findArticulationPoints()
    ap_count = sum(ap)
    leaf_count = g.countLeaves()

    is_biconnected = g.isConnected() and ap_count == 0

    print("\nIs Graph Biconnected:", is_biconnected)
    print("Number of Articulation Points:", ap_count)
    print("Number of Node Leaves:", leaf_count)

File: Graphs/BiconnectedGraph.java
// File Name: BiconnectedGraph.java

import java.util.*;

class BiconnectedGraph {

    private int V;
    private LinkedList<Integer>[] adj;
    private int time = 0;

    BiconnectedGraph(int v) {
        V = v;
        adj = new LinkedList[v];
        for (int i = 0; i < v; i++)
            adj[i] = new LinkedList<>();
    }

    void addEdge(int u, int v) {
        adj[u].add(v);
        adj[v].add(u);
    }

    void APUtil(int u, boolean[] visited, int[] disc,
                int[] low, int[] parent, boolean[] ap) {

        int children = 0;
        visited[u] = true;
        disc[u] = low[u] = ++time;

        for (int v : adj[u]) {

            if (!visited[v]) {
                children++;
                parent[v] = u;
                APUtil(v, visited, disc, low, parent, ap);

                low[u] = Math.min(low[u], low[v]);

                if (parent[u] == -1 && children > 1)
                    ap[u] = true;

                if (parent[u] != -1 && low[v] >= disc[u])
                    ap[u] = true;

            } else if (v != parent[u]) {
                low[u] = Math.min(low[u], disc[v]);
            }
        }
    }

    boolean[] findArticulationPoints() {
        boolean[] visited = new boolean[V];
        int[] disc = new int[V];
        int[] low = new int[V];
        int[] parent = new int[V];
        boolean[] ap = new boolean[V];

        Arrays.fill(parent, -1);

        for (int i = 0; i < V; i++)
            if (!visited[i])
                APUtil(i, visited, disc, low, parent, ap);

        return ap;
    }

    void DFS(int v, boolean[] visited) {
        visited[v] = true;
        for (int n : adj[v])
            if (!visited[n])
                DFS(n, visited);
    }

    boolean isConnected() {
        boolean[] visited = new boolean[V];
        DFS(0, visited);
        for (boolean v : visited)
            if (!v)
                return false;
        return true;
    }

    int countLeaves() {
        int count = 0;
        for (int i = 0; i < V; i++)
            if (adj[i].size() == 1)
                count++;
        return count;
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of vertices: ");
        int V = sc.nextInt();

        System.out.print("Enter number of edges: ");
        int E = sc.nextInt();

        BiconnectedGraph g = new BiconnectedGraph(V);

        System.out.println("Enter edges (u v):");
        for (int i = 0; i < E; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            g.addEdge(u, v);
        }

        boolean[] ap = g.findArticulationPoints();

        int ap_count = 0;
        for (boolean b : ap)
            if (b) ap_count++;

        boolean is_biconnected = g.isConnected() && ap_count == 0;

        System.out.println("\nIs Graph Biconnected: " + is_biconnected);
        System.out.println("Number of Articulation Points: " + ap_count);
        System.out.println("Number of Node Leaves: " + g.countLeaves());

        sc.close();
    }
}
